### ISSUE TITLE
citations: add uniqueness constraint

### ISSUE THREAD

**Issue by gsrohde** [PecanProject/bety#181]
* [ ] _Optional but recommended--see below._ Normalize white space in title and author columns
* [ ] Consolidate duplicate citation rows
* [ ] Add citation uniqueness constraint
* [ ] _Optional--see below._ Add a trigger function to ensure title and author columns are whitespace-normalized upon insertion or update

The minimal constraint we wish to add is

```sql
ALTER TABLE citations ADD UNIQUE (author, year, title);
```

To see violators, run

```sql
SELECT ARRAY_AGG(id), author, year, title FROM citations GROUP BY author, year, title HAVING count(*) > 1;
```

This currently produces 8 pairs of duplicates on ebi_production.

Ideally, however, we would like to go beyond this minimal constraint and prevent the occurrence of two rows where the authors and/or titles don't exactly match but differ only minimally--for example, they differ only in whitespace or capitalization.  To see the problem, run this variant of the above query:

```sql
SELECT
    ARRAY_AGG(id),
    REGEXP_REPLACE(TRIM(LOWER(author)), ' +', ' '),
    year,
    REGEXP_REPLACE(TRIM(LOWER(title)), ' +', ' ')
FROM
    citations
GROUP BY
    REGEXP_REPLACE(TRIM(LOWER(author)), ' +', ' '),
    year,
    REGEXP_REPLACE(TRIM(LOWER(title)), ' +', ' ')
HAVING
    COUNT (*) > 1;
```

This produces 11 pairs of duplicates--three more than when we require exact matching.

We can go even further in this direction by using the PostgreSQL trigram extension:

```SQL
CREATE EXTENSION pg_trgm; -- provides the 'similar' function

SELECT
    c1.id,
    c1.author,
    c1.year,
    c1.title,
    c2.id,
    c2.author,
    c2.year,
    c2.title
FROM
    citations c1
JOIN citations c2 ON (
    c1.id < c2.id
    AND SIMILARITY(c1.author, c2.author) > 0.9
    AND c1.year = c2.year
    AND SIMILARITY(c1.title, c2.title) > 0.5
);
```

This finds a few more potential duplicates (compare citations with ids 12 and 13 and with ids 675 and 676 on ebi_forecast).

# Options for preventing near matches.

Here are two possible sets of constraints to prevent near matches:

```sql
ALTER TABLE citations ADD CONSTRAINT normalized_titles ADD CHECK title = REGEXP_REPLACE(TRIM(title), ' +', ' ')
ALTER TABLE citations ADD CONSTRAINT normalized_authors ADD CHECK author = REGEXP_REPLACE(TRIM(author), ' +', ' ')
ALTER TABLE citations ADD UNIQUE (LOWER(author), year, LOWER(title));
```

Alternatively:

```sql
ALTER TABLE citations ADD UNIQUE REGEXP_REPLACE(TRIM(LOWER(author)), ' +', ' '), year, REGEXP_REPLACE(TRIM(LOWER(title)), ' +', ' ')
```

The first set has the advantage of ensuring the title and author fields are always whitespace-normalized.  This may be a useful thing for client programs to be able to rely upon for display, searching, matching, etc.

The second set has the advantage of requiring minimal cleanup.  But this is a small advantage, since the normalization is easily done:

```sql
UPDATE citations SET title = REGEXP_REPLACE(TRIM(title), ' +', ' ');
UPDATE citations SET author = REGEXP_REPLACE(TRIM(title), ' +', ' ');
```

In either case, a trigger function could also be added to ensure that the author and title columns are automatically normalized when an insertion or update occurs.

**Comment by dlebauer** [comment]
## Fixed citations with null year

* added manually through interface
* removed one with no information (citation_id = 731)
```sql
delete from citations_sites where site_id = 1160 and citation_id = 731;
delete from citations where id = 731;
```
* removed citation only used on orphaned management
```sql
delete from citations_treatments where citation_id = 733;
delete from managements where citation_id = 733;
delete from citations where id = 733;
```

**Comment by gsrohde** [comment]
Constraint
```sql
CREATE UNIQUE INDEX author_year_title ON citations (LOWER(author), year, LOWER(title));
```
has been added to db/migrate/PROPOSED_CONSTRAINTS.

**Comment by dlebauer** [comment]
@gsrohde is this ready to be deployed and the issue closed?

**Comment by gsrohde** [comment]
@dlebauer 

There still two pairs of duplicates because of new citations created in April:
```sql
SELECT ARRAY_AGG(id), author, year, title FROM citations GROUP BY author, year, title HAVING count(*) > 1;
        array_agg        |     author      | year |                                                              title                 
                                             
-------------------------+-----------------+------+------------------------------------------------------------------------------------
---------------------------------------------
 {2000000010,2000000005} | Serbin, Shawn   | 2015 | Remotely estimating photosynthetic capacity, and its response to temperature, in ve
getation canopies using imaging spectroscopy
 {2000000003,842}        | Shawn P. Serbin | 2014 | Spectroscopic determination of leaf morphological and biochemical traits for northe
rn temperate and boreal tree species
(2 rows)
```
Once this is fixed, I can add the uniqueness constraint.

**Comment by dlebauer** [comment]
@serbinsh can you fix these duplicate records on your machine?

**Comment by serbinsh** [comment]
Do I just need to remove one of the records for each?

**Comment by serbinsh** [comment]
Done

**Comment by dlebauer** [comment]
@serbinsh did you make sure to update all referents?

 here is a function for removing duplicates: https://github.com/PecanProject/bety/issues/185 it doesn't work across sites, but this could be solved if db managers for all of the sites run it.

**Comment by serbinsh** [comment]
@dlebauer What do you mean?

I removed both dups:

```
bety=> SELECT ARRAY_AGG(id), author, year, title FROM citations GROUP BY author, year, title HAVING count(*) > 1;
 array_agg | author | year | title
-----------+--------+------+-------
(0 rows)
```

**Comment by dlebauer** [comment]
@serbinsh did you update all of the tables with a `citation_id` field?

**Comment by serbinsh** [comment]
Probably not, can you elaborate? What do you mean "update"? Sorry for being slow I just feel like I missed something

**Comment by gsrohde** [comment]
@serbinsh If you go to http://bety-dev:3000/schemas?partial=citations_table, you can see all the tables that have a `citation_id` column.  (Check Show One Degree and uncheck Show Two Degree to get a simplified view.)  No row in any of these tables should have a value in the `citation_id` column corresponding to the id of a row you deleted.

**Comment by serbinsh** [comment]
@gsrohde I can't access that site.....I get a server not found error

**Comment by gsrohde** [comment]
Sorry, I meant https://www.betydb.org/schemas?partial=citations_table.

**Comment by gsrohde** [comment]
Actually, though, it's only the traits table you have to worry about.  We have foreign-key constaints in place for all of the other tables, so you wouldn't be able to delete the citation if any of those other tables refer to it.

**Comment by serbinsh** [comment]
Yeah, I could not find anything on that schemas page.  I think we are OK.

**Comment by dlebauer** [comment]
@serbinsh you should have seen this:

![image](https://cloud.githubusercontent.com/assets/464871/18930076/47b19d2e-858c-11e6-9677-5f14a35111c7.png)

**Comment by gsrohde** [comment]
Slight correction to my earlier remark: Not only do you have to worry about traits referring to the deleted citations.  It is also the case that if the `citations_sites` or `citations_treatments` table refers to a citation you are deleting, the referring rows will be silently deleted as well.  `managements`, `methods`, `priors`, and `yields` will all complain if you try to delete a referred-to citation, either via SQL or the Web interface.
